dask/config.py

```diff
@@ -484,7 +509,7 @@ def collect(paths: list[str] = paths, env: Mapping[str, str] | None = None) -> d
     if env is None:
         env = os.environ
 
-    configs = collect_yaml(paths=paths)
+    configs: list[dict] = collect_yaml(paths=paths)  # type: ignore
```

Comment:
To prevent this you could have used `@overload` above

Reply:
I'm not sure if I follow. The only thing I can think of is something like checking the mapping items:
`all(isinstance(key, str) and isinstance(value, str) for key, value in configs.items())` 
which I know to be true in this case due to `return_paths=False`.

Reply:
The problem is that you have a Union return type, which is considered bad practice (https://github.com/python/mypy/issues/1693) exactly for this reason.

You can fix it by either changing the return type to Any or by using `@overload` as shown below. I'll leave to you to decide if the extra bulk of `@overload` is justified in this case.

#### Either
```python
@overload
def collect_yaml(
    paths: Sequence[str] = paths,
    *,
    return_paths: Literal[False] = False,
) -> list[dict]:
    ...

@oveload
def collect_yaml(
    paths: Sequence[str] = paths,
    *,
    return_paths: Literal[True],
) -> list[tuple[pathlib.Path, dict]]:
    ...

def collect_yaml(
    paths: Sequence[str] = paths,
    *,
    return_paths: bool = False,
) -> list[dict] | list[tuple[pathlib.Path, dict]]:
    # implementation
```

#### Or
```python
def collect_yaml(
    paths: Sequence[str] = paths,
    *,
    return_paths: bool = False,
) -> Any:
    # implementation
```

Reply:
Holy smokes, first time I've actually used `@overload`, surprising to get C++ vibes when doing python typing. :sweat_smile: 
https://github.com/dask/dask/pull/10930/commits/f44304bc6fbfb1627d30c06ad8c1ea2101232d9d, thanks for the learning!